scripts/lib/CIME/case/case_st_archive.py

```diff
@@ -625,3 +663,79 @@ def case_st_archive(self, last_date_str=None, archive_incomplete_logs=True, copy
+    comp_archive_specs = archive.get_children("comp_archive_spec")
+    for comp_archive_spec in comp_archive_specs:
+        components.append(archive.get(comp_archive_spec, 'compname'))
+        test_file_names = archive.get_optional_child("test_file_names", root=comp_archive_spec)
+        if test_file_names is not None:
+            if not os.path.exists(testdir):
+                os.makedirs(os.path.join(testdir,"archive"))
+
+            for file_node in archive.get_children("tfile", root=test_file_names):
+                fname = os.path.join(testdir,archive.text(file_node))
+                disposition = archive.get(file_node, "disposition")
+                logger.info("Create file {} with disposition {}".
+                            format(fname, disposition))
+                with open(fname, 'w') as fd:
+                    fd.write(disposition+"\n")
+
+    logger.info("testing components: {} ".format(list(set(components))))
+    _archive_process(self, archive, None, False, False,components=list(set(components)),
+                     dout_s_root=dout_s_root,
+                     casename="casename", rundir=testdir, testonly=True)
+
+    _check_disposition(testdir)
+
+    # Now test the restore capability
+    testdir2 = os.path.join(testdir,"run2")
+    os.makedirs(testdir2)
+
+    restore_from_archive(self, rundir=testdir2, dout_s_root=dout_s_root)
+
+    restfiles = [f for f in os.listdir(os.path.join(testdir,"archive","rest","1976-01-01-00000"))]
+    for _file in restfiles:
+        expect(os.path.isfile(os.path.join(testdir2,_file)), "Expected file {} to be restored from rest dir".format(_file))
+
+    return True
+
+def _check_disposition(testdir):
+    copyfilelist = []
+    for root, _, files in os.walk(testdir):
+        for _file in files:
+            with open(os.path.join(root, _file), "r") as fd:
+                disposition = fd.readline()
+            logger.info("Checking testfile {} with disposition {}".format(_file, disposition))
+            if root == testdir:
+                if "move" in disposition:
+                    if find_files(os.path.join(testdir, "archive"), _file):
+                        expect(False,
+                               "Copied file {} to archive with disposition move".format(_file))
+                    else:
+                        expect(False,
+                               "Failed to move file {} to archive".format(_file))
+                if "copy" in disposition:
+                    copyfilelist.append(_file)
+            elif "ignore" in disposition:
+                expect(False, "Moved file {} with dispostion ignore to directory {}".format(_file, root))
+            elif "copy" in disposition:
+                expect(_file in copyfilelist, "File {} with disposition copy was moved to directory {}"
+                       .format(_file, root))
+    for _file in copyfilelist:
+        expect(find_files(os.path.join(testdir,"archive"), _file) != [],
```

Comment:
Now go back through the files in the copylist and make sure that they were copied to the archive.  
With this all of the disposition possibilities and the ways they could go wrong have been covered.